rhombus/private/treelist.rkt

```diff
@@ -0,0 +1,1077 @@
+          (walk (node-ref node bi)
+                other-node
+                si
+                other-index
+                (fx- height 1)
+                other-height)]
+         [else (walk other-node
+                     node
+                     other-index
+                     index
+                     other-height
+                     height)])]
+      [else
+       (define-values (bi si) (if (node-leftwise-dense? node)
+                                  (values (radix index height) index)
+                                  (step node index height)))
+       (define-values (other-bi other-si) (if (node-leftwise-dense? other-node)
+                                              (values (radix other-index height) other-index)
+                                              (step other-node other-index height)))
+       (walk (node-ref node bi)
+             (node-ref other-node other-bi)
+             si
+             other-si
+             (fx- height 1)
+             (fx- height 1))])))
+
+(define (treelist-hash-code tl recur)
+  ;; limit the number of elements that we inspect to 48
+  (define len (treelist-size tl))
+  (cond
+    [(len . fx< . 48)
+     (for/fold ([hc 0]) ([elem (in-treelist tl)])
+       (hash-code-combine hc (recur elem)))]
+    [else
+     (hash-code-combine
+      ;; first 16
+      (for/fold ([hc 0]) ([i (in-range 0 16)])
+        (hash-code-combine hc (recur (treelist-ref tl i))))
+      ;; sparse middle 16
+      (let* ([n (fx- len 32)]
+             [skip (quotient n 16)])
+        (for/fold ([hc 0]) ([i (in-range 16 (fx- len 16) skip)])
+          (hash-code-combine hc (recur (treelist-ref tl i)))))
+      ;; last 16
+      (for/fold ([hc 0]) ([i (in-range (fx- len 16) len)])
+        (hash-code-combine hc (recur (treelist-ref tl i)))))]))
+
+;; functionally update the slot at `index` to `el`
+(define (treelist-set tl index el)
+  (cond
+    [(impersonator? tl)
+     (treelist-set/slow tl index el)]
+    [else
+     (check-treelist 'treelist-set tl)
+     (define size (treelist-size tl))
+     (check-treelist-index 'treelist-set tl size index)
+     (define height (treelist-height tl))
+     (define new-node
+       (let set ([node (treelist-root tl)]
+                 [index index]
+                 [el el]
+                 [height height])
+         (cond
+           [(fx= height 0)
+            (node-set node (radix index height) el)]
+           [(node-leftwise-dense? node)
+            (define branch-index (radix index height))
+            (node*-set node branch-index (set (node*-ref node branch-index) index el (fx- height 1)))]
+           [else
+            (define-values (branch-index subindex) (step node index height))
+            (node-set node branch-index (set (node-ref node branch-index) subindex el (fx- height 1)))])))
+     (treelist new-node size height)]))
+
+;; add `el` to end of vector
+(define (treelist-add tl el)
+  (cond
+    [(impersonator? tl)
```

Comment:
There's a `(fx= size 0)` check as a shortcut a little later. I'm not sure it would be faster to check against `empty-node`, since `empty-node` isn't a compile-time constant, so it has to be fetched at run time.

Reply:
Checking `(eq? tl empty-node)` makes the benchmark slower on my machine.